### PR TITLE
Fix receipt disappears when dismissing distance editor after receipt is generated

### PR DESCRIPTION
**Pull request by wildan-m**
### Details

This PR will fix receipt disappears when dismissing distance editor after receipt is generated

### Fixed Issues

$ https://github.com/Expensify/App/issues/48630
PROPOSAL: https://github.com/Expensify/App/issues/48630#issuecomment-2365034676

### Tests

> [!NOTE]  
> When using a local server, make sure to switch the BE to staging on the Troubleshooting page to access the `GetRouteForBackup` API, which is not yet available in dev.
>
> There are known issues after this PR applied map and some other fields will be flickered for a brief moment
> The issues arise due to discrepancies between the `transactionBackup` and `transaction` data during online updates. Additional BE changes are needed to address this. Related [comment](https://github.com/Expensify/App/pull/51519#discussion_r1817955794)

1. Open app
2. Go to workspace chat.
3. Go offline.
4. Submit a distance expense.
5. Go to transaction thread.
6. Click Distance.
7. Go online.
8. After the receipt in the background is generated, dismiss the distance editor by clicking RHP back button.
- [x] Verify that no errors appear in the JS console

### Offline tests

Same as test

### QA Steps

Same as test
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/c7cddcfa-6b23-48e2-9740-276c517abbba

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/f0263285-c653-47de-a2a9-9af7ba81705a

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/8843e40a-34c9-4668-98da-e21474532522

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/6aaf09af-b968-4146-93fa-3eac7212a99b

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/657dc074-296c-4508-8da4-89e58fb2bf34

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/1dff0eed-77f9-4466-b137-2447d5f3ae3b

</details>

### CODE REVIEW

**Comment by ikevin127** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/f009b187-67c4-4485-919c-daea78644ec2

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/260969fc-dfb3-4faf-8eb1-cbb84a469272


</details>

<details>
<summary>iOS: Native</summary>


https://github.com/user-attachments/assets/ae1a1b6a-be6d-4b9f-a659-15d21ae17ebb



</details>

<details>
<summary>iOS: mWeb Safari</summary>


https://github.com/user-attachments/assets/a3efd018-9b1c-4a1b-9c5a-f8ad117335ef



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>


https://github.com/user-attachments/assets/d03edbbc-77d2-4998-be25-8246c129cb4a



</details>

<details>
<summary>MacOS: Desktop</summary>


https://github.com/user-attachments/assets/fa0b81cd-d951-41f4-9adc-3ed488844a20



</details>

File: src/pages/iou/request/step/IOURequestStepDistance.tsx
Comment (ikevin127) [inline src/pages/iou/request/step/IOURequestStepDistance.tsx:81]: @wildan-m ⚠️ I noticed an issue which happens only when the RHP is dismissed for the first time -> the map component in the report resets / reloads. This doesn't happen if we open Distance RHP again and dismiss for the 2nd time.

https://github.com/user-attachments/assets/cc80f95d-fad1-4be3-9645-5a45423a1af6

I'm pretty sure that this will be reported as regression if we ship it this way, because since we fixed the issue in the sense that the map does show up now when dismissing the RHP, we now have that flicker map reset / reload on first time RHP dismissal.

Please look into this and see if you can figure out a way to not have the report map reset / reload when the RHP is dismissed for the first time and instead keep it the way it looks when it loads for the first time while RHP is still open if possible.

**Review by ikevin127 — CHANGES_REQUESTED** [review]
Besides these two things, the code looks good and I think once the requested changes are addressed we'll be good to Approve -> Merge.

File: src/pages/iou/request/step/IOURequestStepDistance.tsx
Comment (wildan-m) [reply src/pages/iou/request/step/IOURequestStepDistance.tsx:81]: @ikevin127 After further investigation, this issue related to my previous mentioned issue https://github.com/Expensify/App/issues/48630#issuecomment-2438017630 where `transactionBackup` is not exactly the same with unmodified `transaction` data.

For instance, in backup, the pendingFields is not removed, this will cause incorrect result here:

https://github.com/Expensify/App/blob/652d2ffa1b099f9c5a1f0cdce780792097f736ad/src/components/ReportActionItem/MoneyRequestView.tsx#L258-L260

I think the only way to resolve that flicker issue and other fields flicker is to make the backup data exactly the same with unmodified transaction data when they are online, and it will require BE change.

@Gonals is it possible that when we call `GetRouteForBackup` we also get the same unmodified transaction data when they are online?

The missing fields can be seen in previous comment https://github.com/Expensify/App/issues/48630#issuecomment-2438017630

Comment (ikevin127) [reply src/pages/iou/request/step/IOURequestStepDistance.tsx:81]: @wildan-m Thanks for the investigation and explanation, I just wanted to point it out and ask @Gonals whether we should fix it now or in another issue.

Comment (wildan-m) [reply src/pages/iou/request/step/IOURequestStepDistance.tsx:81]: I found another issue where the generated map is not clickable due to pendingFields and pendingAction not being removed. Making 'transactionBackup' identical to unmodified 'transaction' when online is crucial to prevent flickering and an unclickable map, but I'm unsure if this is possible from the backend perspective.

**Comment by ikevin127** [comment]
@wildan-m Please address the requested change in https://github.com/Expensify/App/pull/51519#discussion_r1817946510, after which I will 🟢 **Approve** and defer to **Gonals** regarding issue mentioned in https://github.com/Expensify/App/pull/51519#discussion_r1817955794 and whether or not we cah ship it like this and not get penalty in case it's reported as regression post-merge.

File: src/libs/actions/Transaction.ts
Comment (wildan-m) [reply src/libs/actions/Transaction.ts]: @ikevin127 done

**Review by ikevin127 — APPROVED** [review]
Looks good to me!

**Note**: Failing jest test seems unrelated (timeout).

⚠️ @Gonals Please check out this reported issue / discussion in https://github.com/Expensify/App/pull/51519#discussion_r1817955794 before merging as we need your final decision on whether or not we can ship it in its current state and not get a penalty in case the issue will be reported as regression post-merge.

**Comment by wildan-m** [comment]
@ikevin127 please check this commit https://github.com/Expensify/App/pull/51519/commits/198ccb37cc57dde053e88bd539df1cc367679c79 That change will resolve [map not clickable issue](https://github.com/Expensify/App/pull/51519#discussion_r1818103736).

File: src/pages/iou/request/step/IOURequestStepDistance.tsx
Comment (Gonals) [reply src/pages/iou/request/step/IOURequestStepDistance.tsx:81]: I think this has a different root cause, right? Since it requires BE changes, I think it is a different bug

**Comment by ikevin127** [comment]
@wildan-m Thanks for that additional fix, looks good to me! 🚀 **Approved** once again, and now passing it to @Gonals for final review.

File: src/pages/iou/request/step/IOURequestStepDistance.tsx
Comment (wildan-m) [reply src/pages/iou/request/step/IOURequestStepDistance.tsx:81]: > I think this has a different root cause, right? Since it requires BE changes, I think it is a different bug

@Gonals I agree that flickering is not directly related to the original issue and fixing it is not straightforward.

**Review by Gonals — APPROVED** [review]

**Comment by Gonals** [comment]
No emergency. The linter was complaining about files that were not modified

**Pull request merged**
